File: java/src/main/java/vsdk/toolkit/skin/PatchSet.java
package vsdk.toolkit.skin;

import java.util.ArrayList;
import vsdk.toolkit.common.linealAlgebra.Ray;

public final class PatchSet extends Geometry {
    private ArrayList<Patch> patchList;
    private boolean memoryPoolManaged;

    public PatchSet(ArrayList<Patch> input) {
        super(GeometryClassId.PATCH_SET);
        memoryPoolManaged = false;
        patchList = new ArrayList<>();
        for (int i = 0; input != null && i < input.size(); i++) {
            patchList.add(input.get(i));
        }

        Geometry.patchListBounds(getPatchList(), boundingBox);
        boundingBox.enlargeTinyBit();
        bounded = true;
    }

    @Override
    public void destroy() {
        if (patchList != null) {
            patchList = null;
        }
        super.destroy();
    }

    /**
    DiscretizationIntersect returns null if the ray does not hit the discretization
    of the object. If the ray hits the object, a hit record is returned containing
    information about the intersection point. See geometry.h for more explanation.

    Tests whether the Ray intersect the patches in the list. See geometry.h
    (GeomDiscretizationIntersect()) for more explanation.
    */
    @Override
    public RayHit discretizationIntersect(
        Ray ray,
        float minimumDistance,
        float[] maximumDistance,
        int hitFlags,
        RayHit hitStore) {
        if (!discretizationIntersectPreTest(ray, minimumDistance, maximumDistance)) {
            return null;
        }

        return Geometry.patchListIntersect(patchList, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
    }

    public ArrayList<Patch> getPatchList() {
        return patchList;
    }

    public boolean isMemoryPoolManaged() {
        return memoryPoolManaged;
    }

    public void setMemoryPoolManaged(boolean value) {
        memoryPoolManaged = value;
    }
}

File: typescript/src/main/vsdk/toolkit/skin/PatchSet.ts
import { Ray } from "../common/linealAlgebra/Ray";
import { Geometry } from "./Geometry";
import { GeometryClassId } from "./GeometryClassId";
import type { Patch } from "./Patch";
import type { RayHit } from "./RayHit";

export class PatchSet extends Geometry {
  private patchList: Patch[] | null;
  private memoryPoolManaged: boolean;

  public constructor(input: Patch[] | null) {
    super(GeometryClassId.PATCH_SET);
    this.memoryPoolManaged = false;
    this.patchList = [];
    for (let i = 0; input !== null && i < input.length; i++) {
      this.patchList.push(input[i]);
    }

    Geometry.patchListBounds(this.getPatchList(), this.boundingBox);
    this.boundingBox.enlargeTinyBit();
    this.bounded = true;
  }

  public override destroy(): void {
    if (this.patchList !== null) {
      this.patchList = null;
    }
    super.destroy();
  }

  public override discretizationIntersect(
    ray: Ray,
    minimumDistance: number,
    maximumDistance: number[],
    hitFlags: number,
    hitStore: RayHit | null
  ): RayHit | null {
    if (!this.discretizationIntersectPreTest(ray, minimumDistance, maximumDistance)) {
      return null;
    }

    return Geometry.patchListIntersect(this.patchList, ray, minimumDistance, maximumDistance, hitFlags, hitStore);
  }

  public getPatchList(): Patch[] | null {
    return this.patchList;
  }

  public isMemoryPoolManaged(): boolean {
    return this.memoryPoolManaged;
  }

  public setMemoryPoolManaged(value: boolean): void {
    this.memoryPoolManaged = value;
  }
}
